api/db/crud.py

```diff
@@ -0,0 +1,220 @@
+from tortoise.transactions import in_transaction
+from services.github.client import GithubClient, GithubApiException
+from .models import Session, Player, SourceCode, File, Repository, Comment
+from uuid import UUID
+
+LOGGER = logging.getLogger()
+
+
+class DBCrudException(Exception):
+    pass
+
+
+class PlayerRejoiningError(DBCrudException):
+    pass
+
+
+class NoSelectedCodeError(DBCrudException):
+    pass
+
+
+@asynccontextmanager
+async def __session_lock(session_id: str) -> Session:
+    async with in_transaction():
+        session = await Session.get(id=session_id)
+        await session.select_for_update()
+        yield session
+        session.version += 1
+        await session.save(update_fields=["version"])
+        return
+
+
+async def __load_repos(author_id: UUID, gh_client: GithubClient):
+    author = await Player.get(id=author_id)
+    repo_dicts, _ = await gh_client.get_non_forked_repos(author.username)
+    repos = []
+    for repo_dict in repo_dicts:
+        repos.append(
+            Repository(
+                **repo_dict,
+                author_id=author.id,
+            )
+        )
+
+    await Repository.bulk_create(repos)
+
+
+REPO_LOAD_BATCH_SIZE = 5
+SUPPORTED_LANGUAGE_EXTENSIONS = [
+    "py",
+    "js",
+    "ts",
+    "jsx",
+    "tsx",
+    "go",
+    "dart",
+    "java",
+    "cc",
+    "cpp",
+    "c",
+    "swift",
+]
+
+
+async def __load_files(session_id: str, author_id: UUID, gh_client: GithubClient):
+    async with in_transaction():
+        repos = await Repository.filter(author_id=author_id, is_loaded=False).order_by(
+            "last_pushed_at"
+        )
+        non_empty_repos = 0
+        loaded_repos = []
+        files = []
+        for repo in repos:
+            if non_empty_repos == REPO_LOAD_BATCH_SIZE:
+                break
+            try:
+                file_dicts = await gh_client.get_files_for_repo(
+                    repo.name, repo.default_branch, SUPPORTED_LANGUAGE_EXTENSIONS
+                )
+                for file_dict in file_dicts:
+                    files.append(
+                        File(
+                            **file_dict,
+                            repo_id=repo.id,
+                            author_id=author_id,
+                            session_id=session_id,
+                        )
+                    )
+                if len(file_dicts) > 0:
+                    non_empty_repos += 1
+            except GithubApiException as e:
+                LOGGER.exception(e)
+            finally:
+                loaded_repos.append(repo.id)
+
+        await Repository.filter(id__in=loaded_repos).update(is_loaded=True)
+        await File.bulk_create(files)
+
+
+async def __has_files_to_pick(session_id: str):
+    return await File.filter(session_id=session_id).count() > 0
+
+
+async def __pick_next_file_for_code(session_id: str):
+    authors = (
+        await File.filter(session_id=session_id)
+        .annotate(count=Count("id"))
```

Comment:
What does this do? `Count`

Reply:
Lines 108-112 replicate the following SQL: 

```sql
select author_id, count(distinct id) from file where session_id=SESSION_ID group by author_id
```
`__pick_next_file_for_code`'s purpose is to pick a random file created by an author. In order to make the random selection unweighted with respect to the # of files created per each author, we first randomly select the author, then randomly select a repo corresponding to the author and then randomly select the file.

The above SQL returns all the authors who have a corresponding file in the Session.